src/Controller/ViewerController.php

```diff
@@ -103,4 +119,57 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
     return $response;
   }
 
+  /**
+   * Gets a render array for a cool viewer.
+   *
+   * @param \Drupal\media\Entity\Media $media
+   *   The media entity to view / edit.
+   * @param string $wopi_client
+   *   The WOPI client url.
+   * @param bool $can_write
+   *   Whether this is a viewer (false) or an edit (true). Permissions will
+   *   also be checked.
+   * @param array{closebutton: bool} $options
+   *   Options for the renderer. Current values:
+   *     - "closebutton" if "true" will add a close box. (see COOL SDK)
+   *
+   * @return array|array{error: string}
+   *   A stub render element array, or an array with an error on failure.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The key to use by Collabora is empty or not configured.
+   */
+  protected function getViewerRender(MediaInterface $media, string $wopi_client, bool $can_write, $options = NULL) {
+    $cool_settings = $this->config('collabora_online.settings')->get('cool');
+    $wopi_base = $cool_settings['wopi_base'];
+    $allowfullscreen = $cool_settings['allowfullscreen'] ?? FALSE;
+
+    $id = $media->id();
+
+    $access_token = $this->tokenManager->encode(
+      [
+        'fid' => $id,
+        'uid' => $this->currentUser()->id(),
+        'wri' => $can_write,
+      ],
+      $expire_timestamp,
+    );
+
+    $render_array = [
+      '#wopiClient' => $wopi_client,
+      '#wopiSrc' => urlencode($wopi_base . '/cool/wopi/files/' . $id),
```

Comment:
The only unreliable part would be `$wopi_base`, which ideally we validate in the form.

Note that the `$wopi_base` is a url for the Drupal website, but it can be different from the public url of the Drupal website.

Reply:
Yes, so enough for the moment.